### DISCUSSION TITLE
MQTTNet next major Release

### DISCUSSION THREAD

**Discussion by rido-min** [dotnet/MQTTnet#1923] [General]
As discussed with @chkr1011 offline, we'd like to propose a new major release with the next goals:

- Update the .NET support platforms. Target .NET LTS versions (6, 8) and remove UWP and netstandard. Customers requiring support for the removed platforms can still use the 4.X version.
- Split Client and Server. This will help to evolve (and support) the client and server independently.
- Sign Packages and DLLs with a DNF certificate. #1777 

There might be additional changes in the API, even breaking changes (per SemVer rules wrt Major Releases):

- Swtich to MQTT5 as the default protocol version
- Revisit the MqttConnectingFailedException
- Remove obsolete APIs (such as MqttApplicationMessage.Payload)

Thanks,
Rido

**Comment by kskog** [comment]
What is the timeframe of .NET 8 support?

Why not add support for .NET 8 in a 4.x minor release? Is it a lot of work to validate compatibility ?

**Reply by rido-min** [reply]
.NET8 is already supported (via `net7.0`)

https://github.com/dotnet/MQTTnet/blob/f82d382f23ec8388ed8a33c48b7ede08e2c18622/Source/MQTTnet/MQTTnet.csproj#L13 

.NET 8 LTS dates are described [here ](https://dotnet.microsoft.com/en-us/download/dotnet)

This proposal, for a new major release, is to drop support for old targets, specifically `uap10.0`.

https://github.com/dotnet/MQTTnet/blob/f82d382f23ec8388ed8a33c48b7ede08e2c18622/Source/MQTTnet/MQTTnet.csproj#L15

There are multiple benefits, starting with the reduction of the supported platform matrix, and a simple build system, to avoid issues like #1770

**Reply by kskog** [reply]
Thanks for clearing that up.

**Comment by mregen** [comment]
I would also like to see support for large payloads using ReadOnlySequence buffers. 
Please see discussion/prototype in #1917, #1918 for details. 

Thanks, Martin

**Comment by chkr1011** [comment]
@rido-min Do you have any updates regarding the signing certificate?

**Reply by rido-min** [reply]
sorry, I thought the signing certiticate (and the signing infrastructure) were being tracked in https://github.com/dotnet-foundation/projects/issues/337. 

Let me know if you want to me to chime in.

**Reply by chkr1011** [reply]
I hoped that you got some insights because it is delayed since Dez 15 th. Is there something we can do to move it forward?

**Comment by chkr1011** [comment]
@rido-min Regarding splitting the code into several libraries I have some ideas I want to share with you. Please let me know what you think or if you have another proposal.

### Strategy A
| Library | Description |
|-|-|
| MQTTnet.Core | Contains the protocol level classes like packets, formatter, AsyncEvent etc. |
| MQTTnet.LowLevelClient | Contains only the low-level client implementation |
| MQTTnet.Client | Contains only the "normal" client implementation. |
| MQTTnet.Server | Contains the server implementation. |
| MQTTnet.Server.AspNet | Former "MQTTnet.AspNetCore" library. Contains the same features as right now. |
| MQTTnet.Extensions.Rpc | Same as right now. |
| MQTTnet.Extensions.ManagedClient | Same as right now. I would still call this an extension because users should build their own clients based on the "normal" client or low-level client. This is only a try to put in the "most wanted" behavior in a library. |
| MQTTnet.Extensions.WebSocket4Net | Same as right now. |

Having so many libraries will give us more insights which components are used based on the download statistics from nuget.
Mostly all libraries will reference the library "MQTTnet.Core". 

### Strategy B
| Library | Description |
|-|-|
| MQTTnet | Contains the protocol level classes like packets, formatter, AsyncEvent etc. and the client implementations "low-level" and "normal" |
| MQTTnet.Server | Contains the server implementation. |
| MQTTnet.Server.AspNet | Former "MQTTnet.AspNetCore" library. Contains the same features as right now. |
| MQTTnet.Extensions.Rpc | Same as right now. |
| MQTTnet.Extensions.ManagedClient | Same as right now. I would still call this an extension because users should build their own clients based on the "normal" client or low-level client. This is only a try to put in the "most wanted" behavior in a library. |
| MQTTnet.Extensions.WebSocket4Net | Same as right now. |

Keeping the "MQTTnet" library will make sure that most users will simply upgrade to a new version and do not need to change the nuget package ID in all projects. It will also keep the nuget statistics.

**Reply by rido-min** [reply]
The main reason to _Split client and server_ is to be able to release fixes for each. 

I'd prefer StrategyB, since having multiple packages will introduce additional overhead to manage versioning, and also - as you mentioned - it will be easier to migrate existing projects.

**Comment by chkr1011** [comment]
@rido-min I finally have some time to start with the implementation. After the next release of version 4 (next few days) I will start with the new version.

Summary:
- Apply Strategy B
- Only support .NET 6.0 and .NET 8.0
- Remove WebSocket4Net extension (since it is no longer supported)
- Remove ManagedClient extension (since it has some conceptual issues, lots of bugs and it turns out that lots of people have special needs which are not covered in the extension or require lots of options. Maybe there will be a new version implemented from scratch in the future)
- Change to MQTTv5 by default

After this is done, we can test and discuss the next steps.

Do you have any information when the signing certificate is finally available? I would like to sign version 4 as well.

**Reply by rido-min** [reply]
just pinged the dotnet foundation folks, hope they will get back to us soon.

I'd love to collaborate with you in the next release, are you planning to work in a branch on this repo?

This might be a good opportunity to review some API updates that could introduce breaking changes.

Curious to know more about the ManagedClient conceptual issues.

**Reply by chkr1011** [reply]
I will create a new branch for the next version soon and will let you know.

Regarding the managed client: If you go throw the open issues you will find a lot of issues which are related to the managed client and not yet fixed. A lot of them often complain about a special situation like the need to set a user property while the client sends an ACK packet etc. This cannot be done because it requires exposing a new set of events for every small detail of the QoS flows. Also storing the not sent packages in order to properly support QoS 2 is hard to implement. I already startet a managed client V2 in a separate branch but never finished it. We may talk about this in detail if you want.

**Comment by chkr1011** [comment]
@rido-min the new branch is finally here. It is called "version5"...

**Reply by rido-min** [reply]
what's the best way to provide feedback? PRs on top of that branch? filing new issues or discussions?

**Reply by chkr1011** [reply]
I would prefer PRs on the top of that branch. I am afraid that new issues will be "lost" within the "regular" issues.

**Comment by chkr1011** [comment]
@rido-min Here a quick update:
- Client has now version 5.0.0 set as default
- Server is a dedicated project
- nugets are signed (but I have problems publishing them)
- Client will no longer throw an exception when it received a proper ACK from server
- Removed UWP dedicated code
- Removed .NET 4.5.2 and 4.6 dedicated code
- Removed all things which are marked as obsolete
- Removed WebSocket4Net extension completely
- Only .NET 6.0 (since it is LTS), .NET 7.0 and .NET 8.0 are set as target frameworks

From my notes the only thing which is missing is the rewrite of the payload type which was already done as a draft by @mregen .
If there is more please let me know.

**Reply by rido-min** [reply]
This looks great !! Thanks @chkr1011 

Are you planning to keep the ManagedClient?

One proposal I have is to start separating the client and server tests. But that is another story.

nit about `CleanSession`: to be MQTT5 friendly I think we should rename to `CleanStart`.


If you need help publishing the signed packages, let me know, I can help offline.

**Comment by SeppPenner** [comment]
@chkr1011 Why Net7? It's not LTS and already deprecated as well since 14th of May 2024.

**Reply by chkr1011** [reply]
I am not sure :smile: If it is already out of support I will remove the target framework. 
Please have a look at the branch and let me know if you have more ideas what can/needs to be changed.

**Reply by rido-min** [reply]
I think @SeppPenner is right, .NET7 is out of support since last week https://dotnet.microsoft.com/en-us/download/dotnet, note that .NET6 will go out of support in November 2024, only 6 months away. 

How about targeting only 8? once the v5-preview is out, we can gather feedback from the community.

 If there is not much demand for older frameworks we could remove a few `#if NET7_0_OR_GREATER` that are required for TLS.

**Reply by chkr1011** [reply]
OK then let's try it that way. I will remove .NET 7 stuff.

**Comment by mregen** [comment]
Hi @chkr1011, sorry I am just back from vacation.

> From my notes the only thing which is missing is the rewrite of the payload type which was already done as a draft by @mregen .

are you planning to implement the payload type or do you need another PR on the v5 branch?

**Reply by mregen** [reply]
Hi @chkr1011 and @rido-min, I reopened the PR #2017 against v5 as a brute force change replacing payload type as ReadOnlySequence to get an idea of the impact. Please let me know how to proceed.

**Reply by chkr1011** [reply]
@mregen I started the review and already created some comments. But I still need some more days to have a look.
@rido-min When I am not wrong everything from our list is done after the merge and I can release a beta right?
@SeppPenner We should start removing the documentation for version 3 then, create a dedicated version 4 branch. I created a draft version of a migration path for version 5 in the Wiki (https://github.com/dotnet/MQTTnet/wiki/Upgrading-guide). Please let me know what you think.

**Reply by rido-min** [reply]
The `version5` branch looks great !! 

Before releasing the beta 5, I have some minor comments (nothing blocking to release the beta though)
 
- Add `<AnalysisLevel>latest-recommended</AnalysisLevel>` and clean Messages (to use the latest C# features)
- Versioning with GitVersion (eg, with [nbgv](https://github.com/dotnet/nerdbank.gitversioning))
- Remove ManagedClient
- Review folders/namespaces (so we can use VS->Project->Sync Namespaces)

**Reply by mregen** [reply]
@chkr1011 I addressed your first set of comments

**Reply by chkr1011** [reply]
@rido-min: What does 'and clean Messages' mean? The folders are configured as 'No namespace provider' in Rider so that they are not added to the namespace at all but allow for better file sorting. Isn't there a similar feature in VS?

**Reply by rido-min** [reply]
>  What does 'and clean Messages' mean?

I see some _messages_ running code analysis in VS (in the Error Window). I can submit a PR to clean those.

>  Isn't there a similar feature in VS?

I'm not familiar with Rider, and I dont know if VS has a similar feature. Aligning namespaces with folders is a common practices, to allow finding the .cs file from a call stack without having access to the full source code, it's just a nit.

What is more interesting IMHO is to revisit the Namespaces, now that the server is in a different assembly, we could simplify/unify `MQTTnet`,  `MQTTnet.Client` and others. It will be a breaking change, and this might be the right time to think about those.

**Reply by chkr1011** [reply]
> I see some messages running code analysis in VS (in the Error Window). I can submit a PR to clean those.
Ah okay. I already did that as well because with the changes from @SeppPenner these are now errors.

**Reply by rido-min** [reply]
@chkr1011 are you open to revisit the main Namespaces, now that the server is in a different assembly, we could simplify/unify `MQTTnet`, `MQTTnet.Client` and others. 

It will be a breaking change, and this might be the right time before the next major release.

**Reply by chkr1011** [reply]
@rido-min I reviewed all namespaces. Please have a look and let me know what you think.

**Reply by mregen** [reply]
Hi @chkr1011, I looked at the namespaces and I would be ok. Namespaces are all differentiated at minimum at the first level (e.g MQTTnet.Client) and not necessarily deeper. @rido-min, is this what you suggest to match all namespace with the directory pattern in depth?

**Reply by rido-min** [reply]
I would prioritize a clean public api-surface over source folder structure.

Eg, this line requires 2 usings:

```cs
using MQTTnet;
using MQTTnet.Client;

IMqttClient mqttClient = new MqttClientFactory().CreateMqttClient();
```

Then, when writing the code to connect and publish:

```cs

using MQTTnet;
using MQTTnet.Client;
using MQTTnet.Formatter;
using MQTTnet.Protocol;

await mqttClient.ConnectAsync(new MqttClientOptionsBuilder()
    .WithProtocolVersion(MqttProtocolVersion.V500)
    .Build());

await mqttClient.PublishStringAsync("test", "hello", MqttQualityOfServiceLevel.AtLeastOnce);
```

I have to add another 2 usings.

I know that current IDEs do a good job adding/sorting/removing usings, but I feel it will be easier if the most common types are under a single namespace such as `MQTTnet`.

**Reply by chkr1011** [reply]
@rido-min but that would lead to a huge namespace including hundreds of files, right?

**Reply by rido-min** [reply]
I made a quick experiment by moving most used types to the root namespace, it resulted in 102 files. Note some namespaces still makes sense, such as Formatter, Packets, or Internal.
 
![image](https://github.com/user-attachments/assets/5d96b039-afb6-4a3b-a92b-81c7802378dd)

If having 100 files at the root is a problem we could keep the same folder structure but using the same namespace.

This might be a good opportunity to review public types that can be marked as internal.

**Reply by mregen** [reply]
Hi @rido-min, is your concern just about the usability of the fluent API? I'd prefer structure over usability if I can chose and leave it as is.

**Reply by chkr1011** [reply]
I am not sure if the optimization of using imports is worth losing the file structure. In my opinion we should leave it as it is.

**Reply by rido-min** [reply]
I feel there is something odd between `MQTTnet` and `MQTTnet.Client` namespaces, as they have a circular dependency. 

I'd say that MqttClientFactory, MqttClient and IMqttClient should live in the same namesapce and now that the server is not there, wondering if `MQTTnet.Client` makes sense, so we could get rid of MQTTnet.Client by moving those types to the root.

NDpend does a good job showing namespace dependencies
![image](https://github.com/user-attachments/assets/457a6d24-760d-4571-bf6f-01824efc4321)

**Reply by chkr1011** [reply]
@rido-min Starting with moving "Client" to the root for starters is fine for me.

**Reply by chkr1011** [reply]
The namespaces are now adjusted.

**Reply by mregen** [reply]
Thanks @chkr1011, I started testing the latest version 5 myget...

**Reply by mregen** [reply]
Hi @chkr1011, we have been using the last 5.0.0.1184-beta Nuget since a while and no regressions were found: I think it would be ready for an official release?

**Reply by chkr1011** [reply]
@mregen That sounds great to me. I will first prepare new branches required for version 4 support, including signing of version 4. As soon as this works, I would like to release one official RC first so that other users can also give feedback.

**Reply by mregen** [reply]
Hi @chkr1011 , I noticed a couple of new 5.0.0.0-RC builds on myget, is that already something that should be consumed and tested? 
We are currently happy campers on the 1184-beta, which seems to be removed from myget...
Thanks, Martin

**Reply by chkr1011** [reply]
The rentention policy for myget automatically deletes older nugets. It can only keep around 50 packages in the free version.

**Reply by mregen** [reply]
ok, that explains it. I tested the RC and there are so far no issues.

**Comment by mregen** [comment]
Hi @chkr1011, the PR #2017 is now ready to gather feedback also with the receive code path. 
There are more optimizations possible to eliminate a payload copy in the receive codepath, but I left this out for now. Also array pool allocation do not take advantage yet of the ReadOnlySequence, they are just single allocations from the pool. But you get the idea..
Please let me know if you have any questions or if you would like to discuss on Teams.
Thanks, Martin

**Reply by mregen** [reply]
> Hi @chkr1011, the PR #2017 is now ready to gather feedback also with the receive code path. There are more optimizations possible to eliminate a payload copy in the receive codepath, but I left this out for now. Also array pool allocation do not take advantage yet of the ReadOnlySequence, they are just single allocations from the pool...

--> change of plan, I created a minimal version #2046 as the receive codepath becomes really complicated to improve without breaking changes, so focus is now to get at least the publish improvements in place and look at the rest later...

**Comment by chkr1011** [comment]
I migrated some of my projects to the new version and will let them run a couple of days/weeks. If there are no issues I will release the new version.

@SeppPenner  I will also start updating the Wiki and delete Version 3 stuff etc.

**Reply by chkr1011** [reply]
So far nothing serious has happened. One thing I want to finish before the release is the support of Extended Authentication. The fix is almost completed. So, I expect that the next major version will be released this year.
